### PR TITLE
Add millisecond option to `ConfigDict.ser_json_timedelta`

### LINKED ISSUE #10256: Add a "millisecond" option to ser_json_timedelta config parameter
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

Currently the ConfigDict `ser_json_timedelta` parameter can return data in ISO format or as a second timestamp; it would be very useful if there could be an option that would allow returning a millisecond timestamp instead.

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/concepts/models/#basic-model-usage)
- [X] [Data serialization](https://docs.pydantic.dev/concepts/serialization/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/concepts/json_schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/concepts/dataclasses/)
- [ ] [Model Config](https://docs.pydantic.dev/concepts/config/)
- [ ] [Field Types](https://docs.pydantic.dev/api/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/concepts/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/concepts/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/concepts/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

### PR DESCRIPTION
**Pull request by ollz272**
## Change Summary

Implements https://github.com/pydantic/pydantic/issues/10256 : adds a millisecond option to `ser_json_timedelta`. These are the changes needed in pydantic to enable the changes: see https://github.com/pydantic/pydantic-core/pull/1427 for the corrosponding pydantic core changes.

## Related issue number

fix #10256

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @sydney-runkle

### CODE REVIEW

**Comment by ollz272** [comment]
Note that tests are not *yet* passing, but i've tested the new tests locally against the pydantic core changes and all looks good there. Just needs https://github.com/pydantic/pydantic-core/pull/1427 looked at and merged so this can get in.

File: pydantic/config.py [hunk 1]
```diff
@@ -563,7 +563,7 @@ class Transaction(BaseModel):
     3. Using `'never'` we would have gotten `user=SubUser(hobbies=['scuba diving'], sins=['lying'])`.
     """
 
-    ser_json_timedelta: Literal['iso8601', 'float']
+    ser_json_timedelta: Literal['iso8601', 'float', 'millisecond']
     """
     The format of JSON serialized timedeltas. Accepts the string values of `'iso8601'` and
```
Comment (sydney-runkle) [inline pydantic/config.py:568]: Will need to change the docstring here.

I'm wondering if we should add support for `'second'` as well, and deprecate `'float'` down the line (equivalent to `'second'`...

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Looks good overall, though let's please:
* Update the docs
* Consider adding a `'second'` option, equivalent to `'float'` as mentioned. We could even deprecate `'float'` in v2 with `'second'` as the replacement.

Thanks for your contribution here!!

**Comment by sydney-runkle** [comment]
Maybe `'seconds'` and `'milliseconds'` makes more sense...

**Comment by sydney-runkle** [comment]
Will get back to you re this API change tomorrow morning 👍 - need to get feedback from the team re official decision.

**Comment by ollz272** [comment]
> Will get back to you re this API change tomorrow morning 👍 - need to get feedback from the team re official decision.

Great thank you, changes on the pydantic core side should be straightforward if you wanna do this. I'll update the docs

**Comment by sydney-runkle** [comment]
Verdict on the API - let's go with:

`iso8601`
`seconds_float`
`milliseconds_float`

For now. We'll need to add deprecation info / docs updates for the `float -> seconds_float` logic. This also provides us with the ability to add `seconds_int` and `milliseconds_int` down the line if that's a high enough demand feature request.

Thanks for your help on this - happy to do another review once you've made those changes!

**Comment by ollz272** [comment]
> Verdict on the API - let's go with:
> 
> `iso8601` `seconds_float` `milliseconds_float`
> 
> For now. We'll need to add deprecation info / docs updates for the `float -> seconds_float` logic. This also provides us with the ability to add `seconds_int` and `milliseconds_int` down the line if that's a high enough demand feature request.
> 
> Thanks for your help on this - happy to do another review once you've made those changes!

Would you like me to just add `seconds_int` and `milliseconds_int`? Wouldn't be too much extra work 😄

**Review by ollz272 — COMMENTED** [review]
Left some of my own comments explaining decisions.

**Comment by ollz272** [comment]
@sydney-runkle Hi!

Thanks for the response. I've now:

- Pinned this to my branch of pydantic core so tests can run in the pipeline properly
- Added seconds_float and milliseconds_float options
- Deprecated float, with docs and a warning
- Updated tests

Let me know if theres anything further you want me to do. And if you want me to add the int options, more than happy to just get this done 👍🏻 

Assigning back to you

**Comment by ollz272** [comment]
please review

File: pydantic/config.py [hunk 2]
```diff
@@ -563,13 +563,16 @@ class Transaction(BaseModel):
...
-    `'float'`. Defaults to `'iso8601'`.
+    'float_seconds'`, `'float_milliseconds'` and `'float'`. Defaults to `'iso8601'`.
 
     - `'iso8601'` will serialize timedeltas to ISO 8601 durations.
+    - `'float_seconds'` will serialize timedeltas to the total number of seconds.
+    - `'float_milliseconds'` will serialize timedeltas to the total number of seconds.
     - `'float'` will serialize timedeltas to the total number of seconds.
+        NOTE: `'float' is deprecated in favour of `'float_seconds'`
```
Comment (sydney-runkle) [inline pydantic/config.py:575]: Let's list deprecated in v2.10

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Nice work here! A couple of things:

* Let's fully deprecate `float` support - see comments above.
* Make sure that you're consistent with `seconds_float` vs `float_seconds`. I'd opt for the former.

DH had some good feedback on the `pydantic-core` end of things re multiplication with floats + performance.

Happy to do one more review here after you make the remaining changes.

File: pydantic/_internal/_config.py [hunk 3]
```diff
@@ -169,6 +169,13 @@ def core_config(self, obj: Any) -> core_schema.CoreConfig:
         """
         config = self.config_dict
 
+        if config.get('ser_json_timedelta') == 'float':
+            warnings.warn(
+                'Using a `float` as an option for `ser_json_timedelta` is deprecated, use a `float_seconds` instead.',
```
[the remark this answers is not in the document]
Comment (davidhewitt) [reply pydantic/_internal/_config.py:174]: We should continue to emit this warning, even if we remove "support". We can remove the coercion in v3, I think.

**Comment by ollz272** [comment]
Right, i've pushed changes related to comments made

I think probably not worth reviewing again until the `pydantic-core` changes are finished.

Only thing to note is there are some type checking errors happening, though i think they're caused by https://github.com/pydantic/pydantic-core/pull/1443

**Comment by sydney-runkle** [comment]
@ollz272,

Sounds good, ping me when you're ready for a final review here.

Cc @Viicos re:

> Only thing to note is there are some type checking errors happening, though i think they're caused by https://github.com/pydantic/pydantic-core/pull/1443

File: pyproject.toml [hunk 4]
```diff
@@ -50,9 +50,7 @@ dependencies = [
     'typing-extensions>=4.6.1; python_version < "3.13"',
     'typing-extensions>=4.12.2; python_version >= "3.13"',
     'annotated-types>=0.4.0',
-    "pydantic-core==2.23.1",
-    # See: https://docs.python.org/3/library/zoneinfo.html#data-sources
-    'tzdata; python_version >= "3.9"',
+    "pydantic-core @ git+https://github.com/ollz272/pydantic-core@add-millisecond",
```
[the remark this answers is not in the document]
Comment (sydney-runkle) [reply pyproject.toml:54]: @ollz272, could you revert this change as well as the `pdm.lock` changes? Thanks!

File: pydantic/config.py [hunk 5]
```diff
@@ -562,13 +562,15 @@ class Transaction(BaseModel):
     3. Using `'never'` we would have gotten `user=SubUser(hobbies=['scuba diving'], sins=['lying'])`.
     """
 
-    ser_json_timedelta: Literal['iso8601', 'float']
+    ser_json_timedelta: Literal['iso8601', 'milliseconds_float', 'seconds_float']
```
Comment (sydney-runkle) [inline pydantic/config.py:565]:
```suggestion
    ser_json_timedelta: Literal['iso8601', 'seconds_float', 'milliseconds_float']
```

Nit pick, let's keep the order the same as that of `pydantic-core`

File: pydantic/config.py [hunk 6]
```diff
@@ -562,13 +562,15 @@ class Transaction(BaseModel):
...
     """
     The format of JSON serialized timedeltas. Accepts the string values of `'iso8601'` and
-    `'float'`. Defaults to `'iso8601'`.
+    'seconds_float'`, `'float_milliseconds'` and `'float'`. Defaults to `'iso8601'`.
```
Comment (sydney-runkle) [inline pydantic/config.py:568]: Should read:

```
The format of JSON serialized timedeltas. Accepts the string values of `'iso8601'`, `'seconds_float'`, and `'milliseconds_float'`. Defaults to `'iso8601'`.
```

File: tests/test_json_schema.py [hunk 7]
```diff
@@ -1778,21 +1779,42 @@ class Outer(BaseModel):
...
         ('float', {'duration': {'default': 300.0, 'title': 'Duration', 'type': 'number'}}),
+        ('seconds_float', {'duration': {'default': 300.0, 'title': 'Duration', 'type': 'number'}}),
+        ('milliseconds_float', {'duration': {'default': 300000.0, 'title': 'Duration', 'type': 'number'}}),
         ('iso8601', {'duration': {'default': 'PT5M', 'format': 'duration', 'title': 'Duration', 'type': 'string'}}),
     ],
 )
-def test_model_default_timedelta(ser_json_timedelta: Literal['float', 'iso8601'], properties: typing.Dict[str, Any]):
-    class Model(BaseModel):
-        model_config = ConfigDict(ser_json_timedelta=ser_json_timedelta)
+def test_model_default_timedelta(
```
[the remark this answers is not in the document]
Comment (sydney-runkle) [reply tests/test_json_schema.py:1787]: Yeah, I think having their own tests would be good so that we can avoid the conditional warning checks.

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Looking great - just a few minor change requests, otherwise, we should be ready to merge!

I have some changes that I'd like to make to `pydantic-core` in the next week, so might wait for that to do the minor release with your changes.

**Comment by ollz272** [comment]
> Looking great - just a few minor change requests, otherwise, we should be ready to merge!
> 
> I have some changes that I'd like to make to `pydantic-core` in the next week, so might wait for that to do the minor release with your changes.

Makes sense, i think the only reason the docs ci is failing is because its not on a release of pydantic core. I'll fix up those points now. Happy to then change the version of pydantic core this points at, or for you guys to do it when you're ready for the pydantic core release.

**Comment by sydney-runkle** [comment]
> Happy to then change the version of pydantic core this points at, or for you guys to do it when you're ready for the pydantic core release.

We can do this! If you could revert the `pdm.lock` and version changes though, that'd be great. Thanks!

**Comment by ollz272** [comment]
@sydney-runkle Hi!

Thanks, i've made your requested changes. Tests were still passing when i checked.

I've reverted the pyproject.toml and pdm.lock changes, so should be a simple change when you're ready.

Anything else let me know!

**Review by sydney-runkle — APPROVED** [review]
Amazing work here. Looking forward to this change.

I'll try to get a `pydantic-core` release out this week so that we can push this across the line.

**Comment by sydney-runkle** [comment]
cc @ollz272, all set to merge :)

**Pull request merged**